File: types/world.ts
// Cluster of similar locations
export interface LocationCluster {
  id: string;
  canonicalName: string;
  centroidEmbedding: number[];
}

// What a character knows
export interface KnowledgeEntry {
  id: string;
  content: string;
  acquiredAt: number; // World time tick
  source: 'witnessed' | 'told' | 'inferred';
  sourceCharacterId?: string;
  sourceMessageId?: string; // ID of the chat message that produced this knowledge
}

// Relationship between characters
export interface Relationship {
  characterId: string;
  sentiment: number; // -1 to 1
  description: string;
  lastInteraction: number; // World time tick
}

// Character in the world
export interface Character {
  id: string;
  name: string;
  description: string;
  currentLocationClusterId: string;
  knowledge: KnowledgeEntry[];
  relationships: Relationship[];
  isPlayer: boolean;
  isDiscovered: boolean;
  encounterChance: number; // 0-1, chance of being discovered when entering their location
  goals?: string; // Short-term or long-term motivation
  createdByMessageId?: string; // ID of the chat message that created this character
}

// World time (action-based)
export interface WorldTime {
  tick: number;
  narrativeTime: string; // "Late afternoon", "The next morning"
}

// A message in a conversation
export interface Message {
  id: string;
  conversationId: string;
  role: 'user' | 'assistant' | 'system';
  content: string;
  timestamp: number; // World time tick
  speakerId?: string; // Character ID who said this
}

// A conversation stream
export interface Conversation {
  id: string;
  type: 'main' | 'offscreen';
  locationClusterId: string;
  participantIds: string[];
  messages: Message[];
  isActive: boolean;
  sourceMessageId?: string; // ID of the chat message that triggered this conversation
}

// An event that happened in the world
export interface WorldEvent {
  id: string;
  timestamp: number;
  locationClusterId: string;
  involvedCharacterIds: string[];
  description: string;
  witnessedByIds: string[];
  isOffscreen: boolean;
  sourceMessageId?: string; // ID of the chat message that triggered this event
}

// Initial location for scenario setup
export interface InitialLocation {
  name: string;
  description: string;
}

// Character config for scenario setup
export interface CharacterConfig {
  name: string;
  description: string;
  isPlayer: boolean;
  initialLocationName: string; // Must match a location name
  encounterChance: number;
  goals?: string;
}

// Scenario configuration for starting a new world
export interface ScenarioConfig {
  title: string;
  description: string;
  initialNarrativeTime: string;
  locations: InitialLocation[];
  characters: CharacterConfig[];
  playerStartingLocation: string; // Where the player begins
}

// The entire world state
export interface WorldState {
  id: string;
  scenario: ScenarioConfig;
  time: WorldTime;
  characters: Character[];
  locationClusters: LocationCluster[];
  events: WorldEvent[];
  conversations: Conversation[];
  playerCharacterId: string;
}

File: internal/models/world.go
package models

// Location description with embedding
type Location struct {
	ID             string    `json:"id"`
	RawDescription string    `json:"rawDescription"`
	Embedding      []float64 `json:"embedding"`
	ClusterID      *string   `json:"clusterId"`
}

// LocationCluster is a cluster of similar locations
type LocationCluster struct {
	ID                string    `json:"id"`
	CanonicalName     string    `json:"canonicalName"`
	CentroidEmbedding []float64 `json:"centroidEmbedding"`
}

// KnowledgeEntry represents what a character knows
type KnowledgeEntry struct {
	ID                string `json:"id"`
	Content           string `json:"content"`
	AcquiredAt        int    `json:"acquiredAt"`
	Source            string `json:"source"` // "witnessed", "told", "inferred"
	SourceCharacterID string `json:"sourceCharacterId,omitempty"`
}

// Relationship between characters
type Relationship struct {
	CharacterID     string  `json:"characterId"`
	Sentiment       float64 `json:"sentiment"` // -1 to 1
	Description     string  `json:"description"`
	LastInteraction int     `json:"lastInteraction"`
}

// Character in the world
type Character struct {
	ID                       string           `json:"id"`
	Name                     string           `json:"name"`
	Description              string           `json:"description"`
	CurrentLocationClusterID string           `json:"currentLocationClusterId"`
	Knowledge                []KnowledgeEntry `json:"knowledge"`
	Relationships            []Relationship   `json:"relationships"`
	IsPlayer                 bool             `json:"isPlayer"`
	IsDiscovered             bool             `json:"isDiscovered"`
	EncounterChance          float64          `json:"encounterChance"`
	Goals                    string           `json:"goals,omitempty"`
	CreatedByMessageID       string           `json:"createdByMessageId,omitempty"`
}

// WorldTime tracks action-based time
type WorldTime struct {
	Tick          int    `json:"tick"`
	NarrativeTime string `json:"narrativeTime"`
}

// Message in a conversation
type Message struct {
	ID             string `json:"id"`
	ConversationID string `json:"conversationId"`
	Role           string `json:"role"` // "user", "assistant", "system"
	Content        string `json:"content"`
	Timestamp      int    `json:"timestamp"`
	SpeakerID      string `json:"speakerId,omitempty"`
}

// Conversation stream
type Conversation struct {
	ID                string    `json:"id"`
	Type              string    `json:"type"` // "main", "offscreen"
	LocationClusterID string    `json:"locationClusterId"`
	ParticipantIDs    []string  `json:"participantIds"`
	Messages          []Message `json:"messages"`
	IsActive          bool      `json:"isActive"`
}

// WorldEvent is an event that happened in the world
type WorldEvent struct {
	ID                   string   `json:"id"`
	Timestamp            int      `json:"timestamp"`
	LocationClusterID    string   `json:"locationClusterId"`
	InvolvedCharacterIDs []string `json:"involvedCharacterIds"`
	Description          string   `json:"description"`
	WitnessedByIDs       []string `json:"witnessedByIds"`
	IsOffscreen          bool     `json:"isOffscreen"`
	SourceMessageID      string   `json:"sourceMessageId,omitempty"`
}

// TimeCosts for different actions
var TimeCosts = map[string]int{
	"speak":   1,
	"move":    5,
	"examine": 2,
	"action":  3,
}
